LUCENE-10482 Allow users to create their own DirectoryTaxonomyReaders with empty taxoArrays instead of letting the taxoEpoch decide

gautamworah96 opened the pull request:
https://issues.apache.org/jira/browse/LUCENE-10482

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Lucene:

* https://issues.apache.org/jira/projects/LUCENE

You will need to create an account in Jira in order to create an issue.

The title of the PR should reference the Jira issue number in the form:

* LUCENE-####: <short description of problem or changes>

LUCENE must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Ref: The JIRA issue [description](https://issues.apache.org/jira/browse/LUCENE-10482) provides a nice summary of the motivation and a general overview of the solution.

# Solution

Modify the access of the constructor and another method.

# Tests

The plan here was to add a sample `AlwaysRefreshDirectoryTaxonomyReader` that always refreshed the `DirectoryTaxonomyReader` so as to demonstrate the need for the expert constructor. This class also uses the `getInternalIndexReader` API that was previously unused/untested. 

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

gautamworah96 committed:
Allow users to create their own DirectoryTaxonomyReaders with empty taxoArrays instead of letting the taxoEpoch decide

gautamworah96 committed:
Add test and a sample AlwaysRefreshDirectoryTaxonomyReader

gautamworah96 committed:
Fix gradle check

gautamworah96 commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,95 @@
+import org.apache.lucene.util.IOUtils;
+import org.junit.Test;
+
+public class TestAlwaysRefreshDirectoryTaxonomyReader extends FacetTestCase {
```
This test class feels a bit weird since it was designed to test the expert constructor and an expert `getInternalIndexReader()` API but it ends up introducing a DTR use case class. I could not think of any other way to test this other than the way I thought it would be used in a prod like scenario. Open to ideas here.

gautamworah96 replied:
The next commit does that exactly

mikemccand commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,95 @@
+    DirectoryTaxonomyReader dtr1 = new AlwaysRefreshDirectoryTaxonomyReader(dir);
+    tw.addCategory(new FacetLabel("b"));
+    tw.commit();
+    DirectoryTaxonomyReader dtr2 = dtr1.doOpenIfChanged();
```
I'm a little confused -- wouldn't the ordinary `DTR` also refresh in this case and then show 3 ordinals too?  I.e. this is not really testing the unique capability of `AlwaysRefreshDTR`?

zhaih replied:
+1 with Mike. I think we probably want to directly simulate the case where the taxonomy index are rolling back, which could be a bit complex...

Another idea is we can assert that `taxoArray` is not initialized (maybe by creating a method returning `taxoArray == null`?), because for the cached init of `DTR` the `taxoArray` will be init'd when construction, but for uncached cases the `taxoArray` will remain `null` until being used.

gautamworah96 replied:
> his is not really testing the unique capability of AlwaysRefreshDTR

Yes and that was the cause of my hesitation in the first comment (I had also mentioned this in the first commit javadoc). The next commit adds the correct test case where DTR fails to pass the use case.

> maybe by creating a method returning taxoArray == null?

That was another way but I did not want to add another API to just be able to demonstrate a rare/expert use case.

mikemccand reviewed (commented):
Thanks @gautamworah96!  I like the idea of this change, opening up expert constructor/APIs, but was confused a bit in the test case.  I left a few comments.

gautamworah96 committed:
wip

gautamworah96 commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,95 @@
+
+public class TestAlwaysRefreshDirectoryTaxonomyReader extends FacetTestCase {
+
+  @Test
```
I copied this over from `TestDirectoryTaxonomyReader.java` and that still uses `@Test` annotations.
Will fix

gautamworah96 committed:
try to see if replacing underlying files works

gautamworah96 committed:
some more wip

gautamworah96 committed:
finally working, getting the error we needed

gautamworah96 commented:
Quick update: I am actively working on this issue and (finally!) managed to reproduce the test case that shows the inconsistency caused by reusing `taxoArrays` on an older checkpoint. This update is to just notify reviewers that another revision to the PR is coming. Cleaning up the code and pushing it soon

gautamworah96 committed:
Reproduce test case where we see an AIOOB with a DTR but not with a AlwaysRefreshDTR,

Also separate it out into a different class directly into the Facet package

Change constructor to be protected instead of public to allow subclasses to be able to modify it.

zhaih commented on lucene/facet/src/java/org/apache/lucene/facet/taxonomy/directory/AlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,66 @@
+ * A modified DirectoryTaxonomyReader that always recreates a new {@link
+ * AlwaysRefreshDirectoryTaxonomyReader} instance when {@link
+ * AlwaysRefreshDirectoryTaxonomyReader#doOpenIfChanged()} is called. This enables us to easily go
+ * forward or backward in time by re-computing the ordinal space during each refresh.
```
Maybe add a note/warn here saying this will result in an always `O(#facet_label)` taxonomy array construction time when refresh?

mikemccand replied:
Hmm in the previous revision was this a test-only class?

I'm nervous about making this available in the `facet` jar -- this class should only be used in exceptional cases, while most (normal) cases should use the normal `DTR` that optimizes the very common "refresh forward" case.  Can we maybe make `DTR` smarter to detect when a "roll backwards / revert" situation is happening on refresh?

zhaih replied:
Yeah I guess that's possible, right now we have a single `EPOCH` to represent the version of taxonomy index. And `EPOCH` is increased only when the taxonomy index was recreated. 

I think we might be able to further define `VERSION` and it will be increased at each commit, and reset to 0 when `EPOCH` is bumped. So that then we can inherit an index iff `EPOCH` is the same and `VERSION` is newer?

gautamworah96 replied:
++ I separated it to a different class because in an earlier [comment](https://github.com/apache/lucene/pull/762#discussion_r840008690) we thought that it could be a good idea to directly expose it to our users. Let's keep it as a test class for now.
We can explore the idea of making `DTR smarter to detect when a "roll backwards / revert" situation` in another issue (we will also have to think about how to handle older indexes etc). One thing at a time..
I'll revert it back to a test class

gautamworah96 replied:
Fixed in [09b8b51](https://github.com/apache/lucene/pull/762/commits/09b8b51fe6e21f3d852b29e763eb56034c99bedf)

zhaih commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,95 @@
+   * <p>It does not check whether the private taxoArrays were actually recreated or no. We are
+   * (correctly) hiding away that complexity away from the user.
+   */
+  public void testAlwaysRefreshDirectoryTaxonomyReader() throws IOException {
```
Thank you for adding the test case, it looks awesome!

zhaih commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    mgr.maybeRefresh();
+    IOUtils.close(mgr, dtr1, dr1);
```
Probably that's a bit redundant, but maybe we can assert that we can find the label "a" but not "b"?

gautamworah96 replied:
Added in commit [09b8b51](https://github.com/apache/lucene/pull/762/commits/09b8b51fe6e21f3d852b29e763eb56034c99bedf)

gautamworah96 committed:
docs

gautamworah96 commented on lucene/facet/src/java/org/apache/lucene/facet/taxonomy/directory/DirectoryTaxonomyReader.java:
```diff
@@ -78,10 +78,20 @@
+   * @param taxoArrays taxonomy arrays that store the parent, siblings, children information
    */
-  DirectoryTaxonomyReader(
+  public DirectoryTaxonomyReader(
```
Changed it to protected

gautamworah96 commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,95 @@
+    tw1.close();
+
+    final DirectoryReader dr1 = DirectoryReader.open(dir1);
+    // using a DirectoryTaxonomyReader here will cause the test to fail and throw a AIOOB exception
```
Is there a way in Java to send a class, exception type as a param to this test case and then check the error thrown? So for DTR we could "actually" verify that an AIOOB is thrown instead of me mentioning it here. 


Something like

 ```
// ClassParam is sent as a param to the test case, once as DTR and next time as AlwaysRefreshDTR
final DirectoryTaxonomyReader dtr1 = new ClassParam(dir1);
```

gautamworah96 replied:
I am also okay with just plainly duplicating all the code and creating a new test case. It is important that we have a test case that tests if an AIOOB is thrown. What do other people think?

zhaih replied:
I believe you can use `expectThrows`?
https://lucene.apache.org/core/7_3_1/test-framework/org/apache/lucene/util/LuceneTestCase.html#expectThrows-java.lang.Class-org.apache.lucene.util.LuceneTestCase.ThrowingRunnable-

gautamworah96 replied:
I meant something like

```
void testCase(DTRClass, exceptionType) {
  // dir1 is created here
  final DirectoryTaxonomyReader dtr1 = new DTRClass(dir1);
  // again some logic here
  if (DTRClass instanceOf AlwaysRefreshDTR) {
    mgr.maybeRefresh();
  } else {
    expectThrows(AIOOB, mgr.maybeRefresh());
  }
}
```

zhaih replied:
I guess I would write a generic function like
```
void <T extends Throwable> testCase(Function<Directory, DirectoryTaxonomyReader> dtrProducer, Class<T> exceptionType)
```

gautamworah96 replied:
Added in [09b8b51](https://github.com/apache/lucene/pull/762/commits/09b8b51fe6e21f3d852b29e763eb56034c99bedf)

gautamworah96 committed:
Revert back to test class
Make the test more robust to the test framework,
Allow it to workaround the various chaos edge cases that the test framework throws,
Actually verify that the refreshed taxonomy can only find the older "a"
label and not the newer "b" label

gautamworah96 commented:
Ooof this new commit was quite a journey. The test case sporadically started failing after I added the two use cases (of testing both the DTR and ARDTR). This led me down to debugging and figuring out that the test-framework randomly adds files and folders to empty directories just to chaos test the setup. It then also randomly disables deletes in directories through VirusChecker and WindowsFS. Accounting for these factors finally made the test case work.

We now also check explictly for the older label "a" and ensure that the new commit label "b" can't be found.

mikemccand commented:
> Ooof this new commit was quite a journey. The test case sporadically started failing after I added the two use cases (of testing both the DTR and ARDTR). This led me down to debugging and figuring out that the test-framework randomly adds files and folders to empty directories just to chaos test the setup. It then also randomly disables deletes in directories through VirusChecker and WindowsFS. Accounting for these factors finally made the test case work.
> 
> Directly modifying index files is not easy :) 
> 
> 
> 
> We now also check explictly for the older label "a" and ensure that the new commit label "b" can't be found.

Thanks for persisting @gautamworah96!  Indeed the randomized file system behavior (the occasional `extraN` files!) is exciting when it strikes.  And simulated virus checkers holding files open is also exciting.  Lucene's randomized testing infra is awesome.

mikemccand commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,195 @@
+          }
+          return null;
+        },
+        ArrayIndexOutOfBoundsException.class);
```
Yay!

mikemccand commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,195 @@
+
+    while (dir1.getPendingDeletions().isEmpty() == false) {
+      // make the test more robust to the OS taking more time to actually delete files
+      if (TestUtil.hasVirusChecker(dir1) || TestUtil.hasWindowsFS(dir1)) {
```
You can instead use an `Assume` statement at the top of the test method, and make sure to preserve these comments explaining why this test cannot handle those.  `grep` through existing tests to find examples of other tests doing this?

gautamworah96 replied:
Done. It does not look too neat but I think it is good enough

mikemccand reviewed (approved):
This looks great!  Thanks @gautamworah96.  I left some minor polishing comments.

gautamworah96 committed:
Small fixes, use an Assume statement to check FS, add CHANGES entry

gautamworah96 committed:
Fix conflict

mikemccand commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,202 @@
+    final Path taxoPath1 = createTempDir(String.valueOf(Instant.now()));
+    final Directory dir1 = newFSDirectory(taxoPath1);
+    try {
+      Assume.assumeFalse(
```
if you move this `Assume` statement up to the very start of the test (before opening `Directory` etc.) then you don't need the `try`/`finally` wrapping around it.

gautamworah96 replied:
I'm not sure if i am missing something. I don't see a method that just disables the test from creating `WindowsFS` FS and the functions to check if the `dir` uses `WindowsFS` takes the `dir` as an input?

gautamworah96 replied:
Added an cleaner annotation in [5027eb5](https://github.com/apache/lucene/pull/762/commits/5027eb56ea4d387007a712a9a21d2d06b01c801e)

gautamworah96 committed:
Use an annotation

mikemccand reviewed (approved):
Looks great!  Thanks @gautamworah96!

zhaih commented:
Pushed, could you also open a backport PR? @gautamworah96

uschindler commented:
Hi,
since this commit all Jenkins WIndows builds fail with this error:

```
FAILED:  org.apache.lucene.facet.taxonomy.directory.TestAlwaysRefreshDirectoryTaxonomyReader.testAlwaysRefreshDirectoryTaxonomyReader

Error Message:
java.nio.file.InvalidPathException: Illegal char <:> at index 13: 2022-04-15T20:35:33.995886500Z-001

Stack Trace:
java.nio.file.InvalidPathException: Illegal char <:> at index 13: 2022-04-15T20:35:33.995886500Z-001
	at __randomizedtesting.SeedInfo.seed([D59CA6A5F246F40:34E0EFAFAAB6C033]:0)
	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:232)
	at java.base/java.nio.file.Path.resolve(Path.java:515)
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.mockfile.FilterPath.resolve(FilterPath.java:156))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleTemporaryFilesCleanup.createTempDir(TestRuleTemporaryFilesCleanup.java:284)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleTemporaryFilesCleanup.createTempDir(TestRuleTemporaryFilesCleanup.java:284))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.LuceneTestCase.createTempDir(LuceneTestCase.java:3076)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.LuceneTestCase.createTempDir(LuceneTestCase.java:3076))
	at org.apache.lucene.facet.taxonomy.directory.TestAlwaysRefreshDirectoryTaxonomyReader.testAlwaysRefreshDirectoryTaxonomyReader(TestAlwaysRefreshDirectoryTaxonomyReader.java:56)
	at org.apache.lucene.facet.taxonomy.directory.TestAlwaysRefreshDirectoryTaxonomyReader.testAlwaysRefreshDirectoryTaxonomyReader(TestAlwaysRefreshDirectoryTaxonomyReader.java:128)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1754))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:942))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:978))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:992))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44))
	at [junit@4.13.1/org.junit.rules.RunRules.evaluate(RunRules.java:20)](mailto:junit@4.13.1/org.junit.rules.RunRules.evaluate(RunRules.java:20))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:819)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:819))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:470)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:470))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:951)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:951))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:836)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:836))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:887)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:887))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:898)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:898))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60))
	at [org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)](mailto:org.apache.lucene.test_framework@10.0.0-SNAPSHOT/org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47))
	at [junit@4.13.1/org.junit.rules.RunRules.evaluate(RunRules.java:20)](mailto:junit@4.13.1/org.junit.rules.RunRules.evaluate(RunRules.java:20))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:370))
	at [randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:826)](mailto:randomizedtesting.runner@2.7.6/com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:826))
	at java.base/java.lang.Thread.run(Thread.java:833)
```

uschindler commented on lucene/facet/src/test/org/apache/lucene/facet/taxonomy/directory/TestAlwaysRefreshDirectoryTaxonomyReader.java:
```diff
@@ -0,0 +1,194 @@
+  private <T extends Throwable> void testAlwaysRefreshDirectoryTaxonomyReader(
+      Function<Directory, DirectoryTaxonomyReader> dtrProducer, Class<T> exceptionType)
+      throws IOException {
+    final Path taxoPath1 = createTempDir(String.valueOf(Instant.now()));
```
This breaks on windows, becaue ":" is not a valid character for file names.

uschindler replied:
Why do we need the actual time at all? Every test has its own directory, so theres no need to have time in the file name. createTempDir makes sure to have a unique file name anyways!

gautamworah96 replied:
Apologies for this. I've replied on the JIRA thread as well. I had switched to using Instant.now() during testing because when the single test case was failing midway it was not able to delete the temporary folder. Subsequent runs of the test would fail because a folder with the particular name was already existing. Non reproducibility is also an issue +1.
Submitting a fix right now

gautamworah96 replied:
Created https://github.com/apache/lucene/pull/814
